File: uva online judge/11080 - Place the Guards.py
def read_test_case():
    n, e = [int(i) for i in input().split()]
    adj_list = [[] for i in range(n)]
    for i in range(e):
        u, v = [int(i) for i in input().split()]
        adj_list[u].append(v)
        adj_list[v].append(u)
    return adj_list, n

def bicolor(adj_list, colors, n, start):
    colors[start], cnts = 0, [1, 0]
    stack = [start]
    while len(stack) > 0:
        u = stack.pop()
        for v in adj_list[u]:
            if colors[v] == -1:
                colors[v] = 1 - colors[u]
                cnts[colors[v]] += 1
                stack.append(v)
            elif colors[v] == colors[u]:                
                return -1
    return min(cnts)

if __name__ == '__main__':
    for i in range(int(input())):
        adj_list, n = read_test_case()
        cnt, colors = 0, [-1] * n
        for i in range(n):
            if colors[i] == -1:                
                min_cnt = bicolor(adj_list, colors, n, i)
                if min_cnt == -1:
                    cnt = -1 
                    break
                cnt += min_cnt if min_cnt > 0 else 1
        print(cnt)

File: uva online judge/11080 - Place the Guards.c
#include <stdio.h>
#include <string.h>
#define MAX_NODES 200
#define MIN(a,b) (((a)<(b))?(a):(b))

void read_test_case(int adj_list[MAX_NODES][MAX_NODES + 1], int *n){
    int e, i, u, v;
    scanf("%d %d", n, &e);
    for(i = 0; i < *n; i++)
        adj_list[i][0] = 0;
    for(i = 0; i < e; i++){
        scanf("%d %d", &u, &v);
        adj_list[u][++adj_list[u][0]] = v;
        adj_list[v][++adj_list[v][0]] = u;
    }
}

int bicolor(int adj_list[MAX_NODES][MAX_NODES + 1], int colors[MAX_NODES], int n, int start){
    int stack[MAX_NODES], stack_size = 1;
    int u, v, i, cnts[3] = {0, 1, 0};
    colors[start] = 1;
    stack[0] = start;
    while(stack_size > 0){
        u = stack[--stack_size];
        for(i = 1; i <= adj_list[u][0]; i++){
            v = adj_list[u][i];
            if(colors[v] == 0){
                colors[v] = 3 - colors[u];
                stack[stack_size++] = v;
                cnts[colors[v]]++;
            }else if(colors[v] == colors[u]){
                return -1;
            }
        }
    }
    return MIN(cnts[1], cnts[2]);
}

int main(){
    int t, n, i, min_cnt, cnt;
    int adj_list[MAX_NODES][MAX_NODES + 1], colors[MAX_NODES];
    for(scanf("%d", &t); t > 0; t--){
        read_test_case(adj_list, &n);
        memset(colors, 0, sizeof(int) * n);
        cnt = 0;
        for(i = 0; i < n; i++){
            if(colors[i] == 0){
                min_cnt = bicolor(adj_list, colors, n, i);
                if(min_cnt == -1){
                    cnt = -1;
                    break;
                }
                cnt += min_cnt > 0 ? min_cnt : 1;
            }
        }
        printf("%d\n", cnt);
    }
    return 0;
}